[WALLU] Added pins, serial communication, and motor command parsing

therealhuy opened the pull request:
Currently WIP, added motor command processing (4 characters, 'M' for motor indication, 'F' or 'B' for direction, and 'M_' for max throttle OR '_ _' a number 0-99). Pins defined on schematic added. Organization/separation into files may be needed to reduce clutter.

therealhuy committed:
[WALLU] Added pins, serial communication, and motor command parsing

Currently WIP, added motor command processing (4 characters, 'M' for motor indication, 'F' or 'B' for direction, and 'M_' for max throttle OR '_ _' a number 0-99). Pins defined on schematic added. Organization/separation into files may be needed to reduce clutter.

therealhuy committed:
[WALLU] Added pins, serial communication, and motor command parsing

Currently WIP, added motor command processing (4 characters, 'M' for motor indication, 'F' or 'B' for direction, and 'M_' for max throttle OR '_ _' a number 0-99). Pins defined on schematic added. Organization/separation into files may be needed to reduce clutter.

therealhuy commented on wallu/wallu.ino:
```diff
@@ -0,0 +1,130 @@
+bool direction_change = false;
+
+// WALL-U
+Motor left_motor = Motor(11, A1, A2);
```
parameters will be changed to pin constants defined at beginning of file.

therealhuy committed:
[WALLU] Pinmodes added

Title

therealhuy committed:
[WALLU] Pinmodes added

Title

therealhuy committed:
[WALLU] Implemented battery monitoring

Added battery monitoring implementation. Map function not used due to integer math.

therealhuy committed:
[WALLU] Implemented battery monitoring

Added battery monitoring implementation. Map function not used due to integer math.

therealhuy committed:
[WALLU] Implemented how motor requests will be handled

Title. If an opposite direction request is made, the WALL-U will attempt to STOP in order to be able to swap directions for future opposite direction requests. Otherwise, complete the request normally.

therealhuy committed:
[WALLU] Implemented how motor requests will be handled

Title. If an opposite direction request is made, the WALL-U will attempt to STOP in order to be able to swap directions for future opposite direction requests. Otherwise, complete the request normally.

therealhuy committed:
[WALLU] Added steering angle onto motor control

Angle of steering wheel will adjust how motors will change their speed. Requested angle vs actual angle may be redundant in comparison to requested direction (current angle may not be needed). Added 'stop' direction.
Current motor request implementation at this time:
M Direction AngleDigit1 AngleDigit2 ThrottleDigit1 ThrottleDigit2.

Current implementation will only handle ONE speed (full throttle). Throttle will be implemented in future.

therealhuy committed:
[WALLU] Added steering angle onto motor control

Angle of steering wheel will adjust how motors will change their speed. Requested angle vs actual angle may be redundant in comparison to requested direction (current angle may not be needed). Added 'stop' direction.
Current motor request implementation at this time:
M Direction AngleDigit1 AngleDigit2 ThrottleDigit1 ThrottleDigit2.

Current implementation will only handle ONE speed (full throttle). Throttle will be implemented in future.

therealhuy commented:
Overview of implemented functions as of this comment:

1.  Battery Monitor: Voltage divider readings -> Percentage implementation is complete.
2.  Serial Read-in: Commands are read in through Serial. Requests and divided/separated by semicolons. Implementation is complete
3. Command Parsing: Motor command parsing is complete. Servos to be added.
4. Motor control: Partial implementation. Diagram representing motor control will be attached on this [link](https://drive.google.com/file/d/14ixD98L6QDh-mKdgpq0p23VvyPZ5Fred/view?usp=sharing). The reason for this implementation is to prevent forcing the motor to overcome the force of rolling in the opposite direction. Gear damage may occur in that situation.
5. Pin definition: All pins have been added with the exception of a microphone. All pins are adjusted to their respective modes (OUTPUT vs INPUT). All pins defined at the beginning of the file.

abdullahelqaq commented on wallu/wallu.ino:
```diff
@@ -0,0 +1,202 @@
+#define BATT_MIN 11.0
+#define BATT_MAX 15.0
+#define BATTERY_R1 1000000
+#define BATTERY_R2 200000
```
Let's move all of this to `wallu/consants.h`

therealhuy replied:
constants.h created

abdullahelqaq commented on wallu/wallu.ino:
```diff
@@ -0,0 +1,202 @@
+int steering_angle = 0;
+int requested_steering_angle = -1;
+MotorDirection requested_dir;
+int requested_throttle = 0;
```
We can house all of this under a `MotorController` class.

abdullahelqaq commented on wallu/wallu.ino:
```diff
@@ -0,0 +1,202 @@
+bool direction_change = false;
+
+bool motor_request = false;
+bool cam_servo_request = false;
```
We can house all of this under a SerialComms class.

abdullahelqaq commented on wallu/wallu.ino:
```diff
@@ -0,0 +1,202 @@
+int battery_level = -1;
+unsigned int rpm_right = -1;
+unsigned int rpm_left = -1;
+byte lock_status = 0;    // 1 for locked, 0 for unlocked
```
Same comment -we can make a `HUD` class here. It sounds extra, but it'll help a ton when it comes to debugging and extending functionality. Miscellaneous variables like these are hard to keep track of.

abdullahelqaq commented on wallu/wallu.ino:
```diff
@@ -0,0 +1,202 @@
+  process_command();
+  Serial.println(requested_dir);
+  Serial.println(requested_throttle);
+}
```
This is neat. Can probably have this as a member function of `SerialComms`.

abdullahelqaq commented on wallu/wallu.ino:
```diff
@@ -0,0 +1,202 @@
+int calculate_battery(int read_in) {
+  // Vout = Vs x R2 / R1 + R2
+  // Vout * (R1 + R2) / R2 = Vs
+  float val = analogRead(read_in);
```
We should be using the `Sensor.h` base class here:

```
class VoltageDivider : public Sensor<float, float> {
    void read_sensor() {
        // analogRead & computation here
    }
}
```

abdullahelqaq commented on wallu/wallu.ino:
```diff
@@ -0,0 +1,202 @@
+ 
+void setup() {
+  // Initialization of all Arduino pins
+  pinMode(BATTERY_MONITOR_PIN, INPUT);
+  pinMode(LEFT_RPM_PIN, INPUT);
+  pinMode(RIGHT_RPM_PIN, INPUT);
+  pinMode(MOTOR_R_PWM, OUTPUT);
+  pinMode(MOTOR_R_IN1, OUTPUT);
+  pinMode(MOTOR_R_IN2, OUTPUT);
+  pinMode(MOTOR_R_PWM, OUTPUT);
+  pinMode(MOTOR_L_IN1, OUTPUT);
+  pinMode(MOTOR_L_IN2, OUTPUT);
+  pinMode(LOCK_SERVO_PIN, OUTPUT);
+  pinMode(CAMERA_SERVO_PIN, OUTPUT);
+  pinMode(LOCK_SENSOR_PIN, INPUT);
+  pinMode(RANGE_TRIG_PIN, INPUT); // need to double check
+  pinMode(RANGE_ECHO_PIN, INPUT);
+  pinMode(RED_LED_EN_PIN, OUTPUT);
+  pinMode(GREEN_LED_EN_PIN, OUTPUT);
+  pinMode(BLUE_LED_EN_PIN, OUTPUT);
```
If we have a class for each of these devices, we can move the `pinMode()` calls to the constructor of each class. Reduces clutter here, since we want our main modules to be as high level as possible.

abdullahelqaq reviewed (changes requested):
The actual implementation of everything looks good. I think we should be creating classes for interfaces and devices though to take advantage of modular design. Let me know if you want me to refactor the code into those classes, I can just checkout your branch and push directly to this PR.

therealhuy committed:
[WALLU] Motor control & constants.h

Implemented motor control. Throttle set to 255 by default temporarily. Moved constants to a separate file.

therealhuy committed:
[WALLU] Motor control & constants.h

Implemented motor control. Throttle set to 255 by default temporarily. Moved constants to a separate file.

therealhuy committed:
[WALLU] Added print all data request. Fixed motor request.

Adjusted serial prints to be serial friendly. All data can be printed at once with request "P". Added auxillury get_pwm function. Fixed throttle percentage being used as pwm. Tested motor request processing.

therealhuy committed:
[WALLU] Added print all data request. Fixed motor request.

Adjusted serial prints to be serial friendly. All data can be printed at once with request "P". Added auxillury get_pwm function. Fixed throttle percentage being used as pwm. Tested motor request processing.

therealhuy committed:
[WALLU] Updated motor request parsing to take into account negative angles

Title & minor bug fixes. Addition of other missed data in print_all().

therealhuy committed:
[WALLU] Updated motor request parsing to take into account negative angles

Title & minor bug fixes. Addition of other missed data in print_all().

abdullahelqaq committed:
[mqtt comms] Added MQTT handshake to controller and vision modules (#14)

therealhuy committed:
Rearranged code into appropiate classes, added last known wallu library to work.

therealhuy committed:
fixed merge conflicts

therealhuy committed:
Added ramp coding, LED modification, fixed motor bug

abdullahelqaq commented on wallu/core/constants.h:
```diff
@@ -32,4 +32,4 @@ enum MotorDirection
     FORWARD,
     BACKWARD
-};
+};
```
Plis add a newline.

abdullahelqaq commented on wallu/core/motor/motor_control.cpp:
```diff
@@ -84,4 +84,12 @@ int MotorControl::perform_movement(MotorDirection dir, int throttle, int angle)
+        return 1;
+    else
+        return 0;
 }
```
Plis add newline

abdullahelqaq commented on wallu/wallu.ino:
```diff
@@ -60,6 +61,28 @@ int calculate_battery(int read_in)
   return 100.0 * (raw_voltage - BATT_MIN) / (BATT_MAX - BATT_MIN);
 }
 
+void set_eye_pattern(String pattern)
```
Not saying we should do this for this PR, but it would be cool to have a class for the eyes too.

abdullahelqaq reviewed (approved):
Looks good - just a couple newline characters missing.

therealhuy committed:
added new lines
